File: src/main/kotlin/com/daniel/goncharov/algorithm/playground/dp/LongestIncreasingSubsequence.java
package com.daniel.goncharov.algorithm.playground.dp;

import java.util.List;

public class LongestIncreasingSubsequence {

    public int lis(final List<Integer> numbers) {
        int size = 0;
        int[] memory = new int[numbers.size()];
        for (int i = 0; i < numbers.size(); i++) {
            int currentNumber = numbers.get(i);
            int max = 0;
            for (int j = i - 1; j >= 0; j--) {
                if (numbers.get(j) < currentNumber) {
                    max = Math.max(max, memory[j]);
                }
            }
            memory[i] = max + 1;
            size = Math.max(size, memory[i]);
        }
        return size;
    }
}

File: src/main/java/com/daniel/goncharov/algorithm/playground/leetcode/dp/LongestIncreasingSubsequence.kt
package com.daniel.goncharov.algorithm.playground.leetcode.dp

class LongestIncreasingSubsequence {

    //Goal nlogn. Some kind of a tree? priority queue? Heap

    fun lengthOfLIS(nums: IntArray): Int {
        val cache = MutableList(nums.size) { 1 }
        for (index in 1 until nums.size) {
            cache[index] = maxFrom(index, nums, cache) + 1
        }
        return cache.maxOrNull() ?: 0
    }

    //find max value of cache that is value of nums less than start
    private fun maxFrom(start: Int, array: IntArray, cache: MutableList<Int>): Int {
        var maxFrom = 0
        for (index in start - 1 downTo 0) {
            if (array[index] < array[start]) {
                maxFrom = maxOf(cache[index], maxFrom)
            }
        }
        return maxFrom
    }
}